### ISSUE TITLE
Should we fetch python sources exclusively from pypi?

### ISSUE THREAD

**Issue by healther** [spack/spack#13427]
I first encountered this problem when dealing with the `node-js`-induced `python@2` restriction of `py-jupyter-notebook` (part of the discussion of #10319) and fixed in #13020. The tar-balls under github.com/<project>/archive (github's "Releases" tab) do not necessarily correspond to the tar-balls released via pypi.
In the particular case of `py-jupyter-notebook` there is a significant difference between the "source"-distribution from github and the "prebuild"-distribution from pypi. In particular in the former case the installation has to do the mimifying of the JS and CSS code, for pypi this is already done. In a less dramatic fashion, I noticed that at least for a number of jupyter packages they don't have a version identifier in their github "releases". These seem to only be added in their publishing script.

Long story short: I think it would be a good idea to strongly discourage non-pypi fetch-url's for python packages and maybe even do a concerted effort to move the existing github-based ones to pypi.

**Comment by adamjstewart** [comment]
I agree with this sentiment, and have written some of it in our [PythonPackage build system docs](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github). The only time I've ever _had_ to use GitHub was in https://github.com/pydata/numexpr/issues/340 where the author had accidentally added his personal `site-cfg` to the PyPI tarball. 

The only benefits of downloading from GitHub are:

- `spack checksum` and `spack versions` actually work
- tarball may come with unit tests not included in the PyPI tarball

> do a concerted effort to move the existing github-based ones to pypi.

#13411 will have a few

**Comment by healther** [comment]
> spack checksum and spack versions actually work

If you use the pypi.io thingy (see eg `py-numpy`) both work for me. The downside of course is that you have to construct the url manually, but that is rather straightforward (not sure if scriptable though...)

> tarball may come with unit tests not included in the PyPI tarball

Agreed, but I think that's a minor downside?

**Comment by adamjstewart** [comment]
> both work for me

You mean when you run `spack versions py-numpy` it finds new versions that you can add to the package? That's not the case for me.

**Comment by healther** [comment]
Mh at least with:
```
% git diff
diff --git a/var/spack/repos/builtin/packages/py-numpy/package.py b/var/spack/repos/builtin/packages/py-numpy/package.py
index e1ca99d74..5f419a404 100644
--- a/var/spack/repos/builtin/packages/py-numpy/package.py
+++ b/var/spack/repos/builtin/packages/py-numpy/package.py
@@ -27,7 +27,7 @@ class PyNumpy(PythonPackage):
         'numpy.distutils.command', 'numpy.distutils.fcompiler'
     ]
 
-    version('1.17.3', sha256='a0678793096205a4d784bd99f32803ba8100f639cf3b932dc63b21621390ea7e')
+#    version('1.17.3', sha256='a0678793096205a4d784bd99f32803ba8100f639cf3b932dc63b21621390ea7e')
     version('1.17.2', sha256='73615d3edc84dd7c4aeb212fa3748fb83217e00d201875a47327f55363cef2df')
     version('1.17.1', sha256='f11331530f0eff69a758d62c2461cd98cdc2eae0147279d8fc86e0464eb7e8ca')
     version('1.17.0', sha256='951fefe2fb73f84c620bec4e001e80a80ddaa1b84dce244ded7f1e0cbe0ed34a')
```
I get
```
% spack versions py-numpy
==> Safe versions (already checksummed):
  1.17.2  1.17.0  1.16.4  1.16.2  1.16.0  1.15.3  1.15.1  1.14.6  1.14.4  1.14.2  1.14.0  1.13.1  1.12.1  1.11.3  1.11.1  1.10.4  1.9.2
  1.17.1  1.16.5  1.16.3  1.16.1  1.15.4  1.15.2  1.15.0  1.14.5  1.14.3  1.14.1  1.13.3  1.13.0  1.12.0  1.11.2  1.11.0  1.9.3   1.9.1
==> Remote versions (not yet checksummed):
  1.17.3
% spack checksum py-numpy
==> Found 1 version of py-numpy:
  
  1.17.3  https://pypi.io/packages/source/n/numpy/numpy-1.17.3.zip
==> 
==> How many would you like to checksum? (default is 1, q to abort) 
==> Downloading...
==> Fetching https://pypi.io/packages/source/n/numpy/numpy-1.17.3.zip
##################################################################################################################################################################################################### 100.0%##################################################################################################################################################################################################### 100.0%##################################################################################################################################################################################################### 100.0%
==> Checksummed 1 version of py-numpy

    version('1.17.3', sha256='a0678793096205a4d784bd99f32803ba8100f639cf3b932dc63b21621390ea7e')
```
However it doesn't work for `py-jupyter-notebook`, there I have to explicitly specify a version that I want to checksum

**Comment by adamjstewart** [comment]
I think this issue can be closed. `PythonPackage` now supports `spack versions` and `spack checksum`. Most of our packages have been moved from GitHub to PyPI, and the fact that PyPI is preferred over GitHub is [well documented](https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github). I would welcome a concerted effort to continue to move from GitHub to PyPI for all remaining packages, but I don't think we need an issue to track this.
